core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/ClusterOperationService.java

```diff
@@ -316,55 +320,75 @@ private void handleHealthChange(Set<String> failedNodes, Set<String> newHealthyN
             Map<String, List<String>> autoRecoveryNodesMap, Map<String, InstanceMetaData> autoRecoveryHostMetadata,
-            Map<String, InstanceMetaData> failedHostMetadata) {
+            Map<InstanceMetaData, Optional<String>> failedHostMetadata) {
         try {
             boolean hasAutoRecoverableNodes = !autoRecoveryNodesMap.isEmpty();
             if (hasAutoRecoverableNodes) {
                 flowManager.triggerClusterRepairFlow(cluster.getStack().getId(), autoRecoveryNodesMap, false, false);
-                updateChangedHosts(cluster, autoRecoveryHostMetadata.keySet(), Set.of(SERVICES_HEALTHY), InstanceStatus.WAITING_FOR_REPAIR,
-                        CLUSTER_AUTORECOVERY_REQUESTED);
+                updateChangedHosts(cluster, autoRecoveryHostMetadata.keySet().stream().collect(Collectors.toMap(host -> host, host -> Optional.empty())),
+                        Set.of(SERVICES_HEALTHY), InstanceStatus.WAITING_FOR_REPAIR,
+                        CLUSTER_AUTORECOVERY_REQUESTED_CLUSTER_EVENT, CLUSTER_AUTORECOVERY_REQUESTED_HOST_EVENT);
             }
             if (!failedHostMetadata.isEmpty()) {
-                updateChangedHosts(cluster, failedHostMetadata.keySet(), Set.of(SERVICES_HEALTHY, SERVICES_RUNNING), SERVICES_UNHEALTHY,
-                        CLUSTER_FAILED_NODES_REPORTED);
+                updateChangedHosts(cluster,
+                        failedHostMetadata.entrySet().stream().collect(Collectors.toMap(e -> e.getKey().getDiscoveryFQDN(), e -> e.getValue())),
+                        Set.of(SERVICES_HEALTHY, SERVICES_RUNNING), SERVICES_UNHEALTHY,
+                        CLUSTER_FAILED_NODES_REPORTED_CLUSTER_EVENT, CLUSTER_FAILED_NODES_REPORTED_HOST_EVENT);
             }
             if (!newHealthyNodes.isEmpty()) {
-                updateChangedHosts(cluster, newHealthyNodes, Set.of(SERVICES_UNHEALTHY, SERVICES_RUNNING), SERVICES_HEALTHY,
-                        CLUSTER_RECOVERED_NODES_REPORTED);
+                updateChangedHosts(cluster, newHealthyNodes.stream().collect(Collectors.toMap(host -> host, host -> Optional.empty())),
+                        Set.of(SERVICES_UNHEALTHY, SERVICES_RUNNING), SERVICES_HEALTHY,
+                        CLUSTER_RECOVERED_NODES_REPORTED_CLUSTER_EVENT, CLUSTER_RECOVERED_NODES_REPORTED_HOST_EVENT);
             }
         } catch (TransactionExecutionException e) {
             throw new TransactionRuntimeExecutionException(e);
         }
     }
 
-    private void updateChangedHosts(Cluster cluster, Set<String> hostNames, Set<InstanceStatus> expectedState,
-            InstanceStatus newState, ResourceEvent resourceEvent) throws TransactionService.TransactionExecutionException {
-        String recoveryMessage = cloudbreakMessagesService.getMessage(resourceEvent.getMessage(), hostNames);
+    private void updateChangedHosts(Cluster cluster, Map<String, Optional<String>> hostNamesWithReason, Set<InstanceStatus> expectedState,
```

Comment:
it would be nice if you could do a bit of refactor here and move out some parts of this method